lib/Transforms/Instrumentation/ShadowMemSeaDsa.cc

```diff
@@ -877,6 +911,227 @@ void ShadowDsaImpl::updateReadMod(Function &F, NodeSet &readSet,
+                                       << ", " << memUse << ")?\n");
+
+  // Right now memUse must be a load instruction.
+  if (isMemInit(memDef))
+    return true;
+
+  // Get the corresponding pointers and check if these two pointers have
+  // disjoint allocation sites -- if not, they cannot alias.
+  // This can be extended further with offset tracking.
+
+  auto lit = m_shadowLoadToLoad.find(&memUse);
+  if (lit == m_shadowLoadToLoad.end())
+    return true;
+
+  auto getTypeSize = [this] (Type *ty) {
+    unsigned bits = m_dl->getTypeSizeInBits(ty);
+    return bits < 8 ? 1 : bits / 8;
+  };
+
+  LoadInst *concreteLoad = lit->second;
+  assert(concreteLoad);
+  assert(memUse.getNextNode() == concreteLoad);
+  Value *loadSrc = concreteLoad->getPointerOperand();
+  const unsigned loadedBytes = getTypeSize(concreteLoad->getType());
+
+  auto sit = m_shadowStoreToStore.find(&memDef);
+  if (sit == m_shadowStoreToStore.end())
+    return true;
+
+  StoreInst *concreteStore = sit->second;
+  assert(concreteStore);
+  assert(memDef.getNextNode() == concreteStore);
+  Value *storeDst = concreteStore->getPointerOperand();
+  const unsigned storedBytes =
+      getTypeSize(concreteStore->getValueOperand()->getType());
+
+  LOG("shadow_optimizer",
+      llvm::errs() << "alias(" << *loadSrc << ", " << *storeDst << ")?\n");
+
+  // If the offsets don't overlap, no clobbering may happen.
+  {
+    assert(m_graph->hasCell(*loadSrc));
+    assert(m_graph->hasCell(*storeDst));
+    const dsa::Cell &loadCell = m_graph->getCell(*loadSrc);
+    const dsa::Cell &storeCell = m_graph->getCell(*storeDst);
+    assert(loadCell.getNode() == storeCell.getNode());
+
+    // This works with arrays in SeaDsa because all array elements share the
+    // same abstract field.
+    const unsigned loadStartOffset = loadCell.getOffset();
+    const unsigned loadEndOffset = loadStartOffset + loadedBytes;
+    const unsigned storeStartOffset = storeCell.getOffset();
+    const unsigned storeEndOffset = storeStartOffset + storedBytes;
+
+    LOG("shadow_optimizer",
+        llvm::errs() << "\tmayClobber[3]\n\tloadStart: " << loadStartOffset
+                     << ", loadEnd: " << loadEndOffset
+                     << "\n\tstoreStart: " << storeStartOffset
+                     << ", storeEnd: " << storeEndOffset << "\n");
+
+    if (loadEndOffset <= storeStartOffset || loadStartOffset >= storeEndOffset)
+      return false;
+  }
+
+  auto useAllocSites = getAllAllocSites(*loadSrc, cache);
+  if (!useAllocSites.hasValue())
+    return true;
+
+  LOG("shadow_optimizer", llvm::errs() << "\tmayClobber[4]\n");
+
+  auto defAllocSites = getAllAllocSites(*storeDst, cache);
+  if (!defAllocSites.hasValue())
+    return true;
+
+  LOG("shadow_optimizer", llvm::errs() << "\tmayClobber[5]\n");
+
+  for (auto *as : *useAllocSites)
+    if (defAllocSites->count(as) > 0)
+      return true;
+
+  LOG("shadow_optimizer", llvm::errs() << "\tmayClobber[6]\n");
+
```

Comment:
Should also use BasicAA and TBAA. In particular, TBAA might be effective since it has extra information from the source language